### ISSUE TITLE
`redundant-expr` false positive with walrus operator

### ISSUE THREAD

**Issue by injust** [python/mypy#19915]
**Bug Report**

When a walrus operator is used on the left side of an `and` expression, mypy thinks it's redundant, even if the variable is used on the right side.

**To Reproduce**

```python
print([y for x in range(5) if (div_mod := divmod(x, 2)) and (y := sum(div_mod))])
```

**Expected Behavior**

No errors

**Actual Behavior**

```
main.py:1: error: Left operand of "and" is always true  [redundant-expr]
```

**Your Environment**

- Mypy version used: 1.18.2
- Mypy command-line flags: `--enable-error-code=redundant-expr`
- Python version used: 3.13

**Comment by sterliakov** [comment]
I'd argue that this is not a false positive - such chaining with `or`/`and` solely for side effects is somewhere at the edge of acceptable coding style IMO. Mypy is correct here: the LHS of this `and` is always true (as a 2-tuple), so the diagnostic isn't wrong. I'm personally all for keeping this behaviour as-is. If you need this `div_mod` value several times, consider assigning it at the first use site instead of an `and` chain:

```python
print([
    y + z
    for x in range(5) 
    if (y := sum(div_mod := divmod(x, 2))) and (z := div_mod[0] * 2)
])
```

**Comment by injust** [comment]
This is a contrived example, but I think walrus operators in the conditional portion of a comprehension are a very valid use case.

Although the error message _is_ correct (the left side is always truthy), the error code is `redundant-expr`, and it definitely isn't redundant because it can't be simply removed.

I'd argue that your snippet has more complicated logic and is actually _less_ readable.

**Comment by A5rocks** [comment]
I feel like this is a rabbit hole better left unexplored. For instance, what if a function mutates some global state that another function needs? Generally "does this expression have side-effects" is something mypy isn't equipped to answer. IMO it's worth the false positives (if you can call them that!) here.

Maybe it's worth special casing walruses like you say, but IMO only if this would actually end up used. This seems like a very uncommon pattern? I think we might just have to admit that this is a case where mypy is more of a linter than a type-checker!

(concretely, this seems like a pattern where adding it to mypy would be more effort than any usage of this adding a type-ignore or rewriting! and I can think of scenarios where this might be useful! (e.g. if someone doesn't realize the walrus is truthy))

**Comment by A5rocks** [comment]
Alright I thought more and I think this isn't worth implementing. The same pattern might merit a `redundant-expr` based on the context and this isn't really something we can do anything about. For instance, if I wrote

```py
if a := f() and b := g(a):
  pass
```

I think I would want to be warned if suddenly `f` can never return a falsey result and rewrite the `if` condition to be `if b := g(f())`. However, if it's something like:
```py
if a := f() and b := g(a):
  print(a)
```
then... I would probably want to write it like:
```py
a = f()
if b := g(a):
  print(a)
```
but if I were in an expression context (like your initial list comprehension) then I'm not sure, either what you originally had or what @sterliakov proposed... I think we should just always warn and people can add a `# type: ignore` if they are wedded to having a side-effect causing expression returning `True`. IMO this is useful enough to merit the slight readability hit for (a) expression contexts that (b) are annotated as always being truthy but (c) are used later in the expression and where (d) the developer prefers `a := f() and b := g(a)` over `b := g(a := f())`.
